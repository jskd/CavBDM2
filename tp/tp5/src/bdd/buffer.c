/**
* TP n°: 3
*
* Titre du TP : Merge Join Duplicate
*
* Date : 21/10/17
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques :
*/
// Pour getline (standard GNU)
#define _GNU_SOURCE

#include "buffer.h"
#include "hexdump.h"
#include "quicksort.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/**
 * @brief Structure de donnée repésentant un buffer
 */
struct buf {
  /// Value: Valeur
  char* v;
  /// Size: Taille
  size_t s;
  /// Count: Nombre de valeur entrée
  size_t c;
};

/**
 * Creation d'un buffer
 * @param[in] size Taille du buffer
 */
struct buf* buf_create( size_t size ) {
  struct buf* buf= (struct buf*) malloc(sizeof(struct buf));
  buf->v= (char*) malloc(size);

  buf->s= size;
  buf->c=0;
  return buf;
}

/**
 * Détruit le buffer
 * @param[in] buf buffer à détruire
 */
void buf_destroy(struct buf* buf) {
  free(buf->v);
  free(buf);
}

/**
 * Ajoute un caractere dans le buffer s'il reste de la place
 * @param[in] buf   buffer d'entrée
 * @param[in] value valeur à entrer
 * @return    0   succès
 *            -1  erreur: buffer plein
 */
char buf_put(struct buf* buf, char value) {
  if(buf->c >= buf->s)
    return -1;

  buf->v[buf->c]= value;
  buf->c++;

  return 0;
}

/**
 * @brief Ouvre un fichier
 *        Stoque son contenu dans un buffer
 *        Ferme le fichier
 *
 * @param[in]  File_name Fichier
 * @param[in] buffer_size Taille du buffer,
 *             si trop petit pour contenir le fichier entirerement
 *             alors le reste du fichier est ignoré
 *
 * @return buffer si NULL alors erreur de lecture du fichier
 *
 */
struct buf* storeFileBuffer(const char* file_name, size_t buffer_size) {
  FILE* file = fopen(file_name, "r");
  if (file == NULL)
    return NULL;

  struct buf* buf= buf_create( buffer_size );

  char* line = NULL;
  size_t len = 0;

  while((getline(&line, &len, file)) != -1) {
    buf_put(buf, line[0]);
  }

  fclose(file);

  return buf;
}

/**
 * @brief Ouvre un fichier
 *        Ecrit le contenu du buffer dans le fichier
 *        Ferme le fichier
 *
 * @param[in]  File_name Fichier
 * @param[in]  buf       Buffer à écrire dans le fichier
 *
 * @return -1 si erreur dans l'ouverture du fichier
 *
 */
char writeBufferInFile(const char* file_name, const struct buf* buf) {

  FILE* file = fopen( file_name, "w");
  if (file == NULL)
    return -1;

  for(int i=0; i<buf->c; i++)
    fprintf(file, "%c\n", buf->v[i]);

  fclose(file);
  return 0;
}

/**
 * @brief Quick sort buffer
 * @param[in] buf buffer à trier
 */
void buf_quicksort(struct buf* buf) {
  quicksort(buf->v, 0, (buf->c) -1);
}

/**
 * @brief Hex dump buffer
 * @param[in] buf buffer à trier
 */
void buf_dump(const struct buf* buf) {
  printf ("buffer count: %zu\n", buf->c);
  printf ("buffer size: %zu\n", buf->s);
  hexDump("buffer value:", buf->v, buf->c);
}

/**
 * @brief Retourne la valeur à la position d'un buffer
 *
 * @param[in]  buf   Buffer
 * @param[in]  index Index of value
 *
 * @return Value
 *         Si -1: Erreur index > count
 */
char buf_val(const struct buf* buf, int index)
{
  if(index < buf->c)
    return buf->v[index];
  return -1;
}

/**
 * @brief Retourne le nombre de caractere enregistré dans le buffer
 *
 * @param[in]  buf       Buffer
 *
 * @return Nombre de caractere enregistré
 */
size_t buf_count(const struct buf* buf)
{
  return buf->c;
}
