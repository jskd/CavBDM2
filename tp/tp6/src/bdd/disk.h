/**
* TP n°: 5
*
* Titre du TP : Nested loop join disk
*
* Date : 10 Nov 2017
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques :
*/

#ifndef SRC_BDD_DISK_H
#define SRC_BDD_DISK_H

#include <stdio.h>
#include "bufferExtended.h"
#include "table.h"

/**
 * @brief Structure de donnée repésentant un buffer
 * Remarque: Structure caché! Ne pas initilialiser autrement
 * qu'avec les fonctions prévu à cet effet!
 */
struct disk;

size_t disk_count(const struct disk* disk);

struct disk* disk_create( const char * dir);

void disk_destroy(struct disk* disk);

FILE* disk_item(const struct disk* disk, int index);

void disk_storeContentInTable(const struct disk* disk, struct buffer* buf, struct table* tab);

#endif
