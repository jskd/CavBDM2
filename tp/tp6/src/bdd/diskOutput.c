/**
* TP n°: 5
*
* Titre du TP : Nested loop join disk
*
* Date : 10 Nov 2017
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques :
*/

#include "diskOutput.h"
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/types.h>

struct disk_output {
  char* dir;
  char* prefix;
  char* extension;
  int file_number;
  FILE* current_file;
};

struct disk_output* disk_output_create(const char* dir, const char* prefix,
  const char* extension, int offset)
{
  struct disk_output* disk_o= (struct disk_output*) malloc(sizeof(struct disk_output));
  disk_o->dir= strdup(dir);
  disk_o->prefix= strdup(prefix);
  disk_o->extension= strdup(extension);
  disk_o->file_number= offset;
  mkdir(dir, 0777);


  char filename[PATH_MAX];
  sprintf(filename, "%s/%s%d%s", disk_o->dir, disk_o->prefix, disk_o->file_number, disk_o->extension);
  disk_o->current_file= fopen(filename, "w+");
  return disk_o;
}

FILE* disk_output_get_current_file_descriptor( struct disk_output* disk_o) {
  return disk_o->current_file;
}

FILE* disk_output_next_file( struct disk_output* disk_o) {

  fclose(disk_o->current_file);

  disk_o->file_number++;
  char filename[PATH_MAX];
  sprintf(filename, "%s/%s%d%s", disk_o->dir, disk_o->prefix, disk_o->file_number, disk_o->extension);
  disk_o->current_file= fopen(filename, "w+");

  return disk_o->current_file;
}


void disk_output_destroy( struct disk_output* disk_o ) {
  free(disk_o->dir);
  free(disk_o->prefix);
  free(disk_o->extension);
  free(disk_o);
}
