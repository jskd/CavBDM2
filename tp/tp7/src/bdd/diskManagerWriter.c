/**
* TP n°: 5
*
* Titre du TP : Nested loop join disk
*
* Date : 10 Nov 2017
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques :
*/

#include "diskWriter.h"
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/types.h>


/**
* TP n°: 5
*
* Titre du TP : Nested loop join disk
*
* Date : 10 Nov 2017
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques :
*/
#include "diskWriter.h"
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <stdio.h>
#include <ftw.h>
#include <unistd.h>
#include "rmrf.h"

/**
 * @brief Structure caché des données repésentant un diskManagerWriter
 */
struct diskManagerWriter{
  char* dir;
  char* d_prefixe;
  char* d_suffixe;
  char* f_prefixe;
  char* f_suffixe;
  int dw_number;
  struct diskWriter* current_dw;
};

static struct diskWriter* _dmw_disk_w_create_from_index(struct diskManagerWriter* dmw, int index) {
  char diskname[PATH_MAX];
  sprintf(diskname, "%s/%s%03d%s", dmw->dir, dmw->d_prefixe, index, dmw->d_suffixe);
  return disk_w_create(diskname, dmw->f_prefixe, dmw->f_suffixe);
}

/**
 * @brief     Creation d'un diskManagerWriter
 *         Les disk écrit sont dans le format "{dst} / {prefix} + {index} + {suffixe}".
 * @param[in] dst      Repertoire de destination
 * @param[in] prefixe  Prefixe de nommage des disks
 * @param[in] suffixe  Sufixe de nommage des disks
 */
struct diskManagerWriter* disk_manager_w_create(const char* dir, const char* disk_prefixe,
  const char* disk_suffixe, const char* file_prefixe, const char* file_suffixe)
{
  struct diskManagerWriter* dmw= (struct diskManagerWriter*) malloc(sizeof(struct diskManagerWriter));
  dmw->dir= strdup(dir);
  dmw->d_prefixe= strdup(disk_prefixe);
  dmw->d_suffixe= strdup(disk_suffixe);
  dmw->f_prefixe= strdup(file_prefixe);
  dmw->f_suffixe= strdup(file_suffixe);

  rmrf(dir);
  mkdir(dir, 0777);

  dmw->current_dw= _dmw_disk_w_create_from_index(dmw, 0);
  dmw->dw_number=1;

  return dmw;
}

/**
 * @brief     Retourne le diskWriter (écrase le contenu s'il existe déjà)
 * @param[in] dmw        DiskManagerWriter
 * @return    diskWriter
 */
struct diskWriter* disk_manager_w_get_current_dw(struct diskManagerWriter* dmw) {
  return dmw->current_dw;
}

/**
 * @brief     Passage au diskWriter suivant
 * @param[in] dmw      DiskManagerWriter
 */
void disk_manager_w_next_dw(struct diskManagerWriter* dmw) {
  disk_w_destroy(dmw->current_dw);
  dmw->current_dw= _dmw_disk_w_create_from_index(dmw, dmw->dw_number);
  dmw->dw_number++;
}

int disk_manager_w_count(struct diskManagerWriter* dmw) {
  return dmw->dw_number;
}

/**
 * @brief     Detruit le diskWriterManger
 * @param[in] dmw        DiskManagerWriter
 */
void disk_manager_w_destroy(struct diskManagerWriter* dmw) {
  disk_w_destroy(dmw->current_dw);
  free(dmw->dir);
  free(dmw->d_prefixe);
  free(dmw->d_suffixe);
  free(dmw->f_prefixe);
  free(dmw->f_suffixe);
  free(dmw);
}
