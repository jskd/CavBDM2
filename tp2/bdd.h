/**
* TP n°: 2
*
* Titre du TP : Merge Join
*
* Date : 13/10/17
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques :
*
* History: 146488d PASS (avec struct)
*          7d1b9de PASS (sans struct)
*/
#include <stdio.h>

/**
 * @brief Structure de donnée repésentant un buffer
 * Remarque: Structure caché! Ne pas initilialiser autrement
 * qu'avec les fonctions prévu à cet effet!
 */
struct buf;

/**
 * Creation d'un buffer
 * @param[in] size Taille du buffer
 */
struct buf* buf_create( size_t size );

/**
 * Détruit le buffer
 * @param[in] buf buffer à détruire
 */
void buf_destroy(struct buf* buf);

/**
 * Ajoute un caractere dans le buffer s'il reste de la place
 * @param[in] buf   buffer d'entrée
 * @param[in] value valeur à entrer
 * @return    0   succès
 *            -1  erreur: buffer plein
 */
char buf_put(struct buf* buf, char value);

/**
 * @brief Enregistre un le 1er caractére de chaque
 *        ligne d'un fichier dans un buffer
 *
 * @param[in]  Fichier d'entrée
 * @param[out] Buffer de sortie
 */
void storeFileBuffer(FILE* fp, struct buf* buf);

/**
 * Ecrit un buffer dans un fichier
 *
 * @param[out] fp  fichier de sortie
 * @param[in]  buf fichier d'entrée
 */
void writeBufferInFile(FILE* fp, struct buf* buf);

/**
 * Natual join
 *
 * @param[in]  buf_a   relation externe
 * @param[in]  buf_b   relation interne
 * @param[out] buf_out resultat du natural join de buf_a et buf_b
 */
void natural_join(struct buf* buf_a, struct buf* buf_b, struct buf* buf_out);

/**
 * @brief Merge join
 *
 * @param[in]  buf_a   relation externe
 * @param[in]  buf_b   relation interne
 * @param[out] buf_out resultat du merge_join
 */
void merge_join(struct buf* buf_a, struct buf* buf_b, struct buf* buf_out);
