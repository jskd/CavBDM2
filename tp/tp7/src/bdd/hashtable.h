/**
* TP n°: 4
*
* Titre du TP : Hash Join
*
* Date : 27/10/17
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques :
*/
#ifndef SRC_BDD_HASHTABLE_H
#define SRC_BDD_HASHTABLE_H

extern const char VAL_NOT_IN_HASHTABLE;

#include <stdio.h>
#include "buffer.h"
/**
 * @brief Structure de donnée repésentant une table de hash
 * Remarque: Structure caché! Ne pas initilialiser autrement
 * qu'avec les fonctions prévu à cet effet!
 */
struct hashtable;

/**
 * Creation d'une hashtable
 * @param[in] m Nombre de bucket
 */
struct hashtable* hashtable_create( size_t m );

/**
 * Suppression d'une hashtable
 * @param[in] hashtable
 */
void hashtable_destroy( struct hashtable* ht );

char hashtable_is_full(struct hashtable* ht);

char hashtable_put(struct hashtable* ht, char key, char val);

char hashtable_get(const struct hashtable* ht, char key);

void hashtable_remove(struct hashtable* ht, char key);

void hashtable_print(struct hashtable* ht);

char storeBufferToHashtable(struct buffer* buf, struct hashtable* ht);

/**
 * Hash fonction (linear)
 * @param[in] ht table de hash cible
 * @param[in] key clef
 */
int hash(const struct hashtable* ht,char key);

struct hashtable* storeFileInHashtable(const char* filename, size_t m);

#endif
