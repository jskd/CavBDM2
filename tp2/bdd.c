/**
* TP n°: 2
*
* Titre du TP : Merge Join
*
* Date : 13/10/17
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques :
*
* History: 146488d PASS (avec struct)
*          7d1b9de PASS (sans struct)
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "bdd.h"

/**
 * @brief Structure de donnée repésentant un buffer
 */
struct buf {
  char* v;  //< Valeur
  size_t s; //< Taille
  size_t c; //< Nombre de valeur entrée
};

/**
 * Creation d'un buffer
 * @param[in] size Taille du buffer
 */
struct buf* buf_create( size_t size ) {
  struct buf* buf= (struct buf*) malloc(sizeof(struct buf));
  buf->v= (char*) malloc(size);
  buf->s= size;
  buf->c=0;
  return buf;
}

/**
 * Détruit le buffer
 * @param[in] buf buffer à détruire
 */
void buf_destroy(struct buf* buf) {
  free(buf->v);
  free(buf);
}

/**
 * Ajoute un caractere dans le buffer s'il reste de la place
 * @param[in] buf   buffer d'entrée
 * @param[in] value valeur à entrer
 * @return    0   succès
 *            -1  erreur: buffer plein
 */
char buf_put(struct buf* buf, char value) {
  if(buf->c >= buf->s)
    return -1;

  buf->v[buf->c]= value;
  buf->c++;

  return 0;
}

/**
 * @brief Enregistre un le 1er caractére de chaque
 *        ligne d'un fichier dans un buffer
 *
 * @param[in]  Fichier d'entrée
 * @param[out] Buffer de sortie
 */
void storeFileBuffer(FILE* fp, struct buf* buf) {
  char* line = NULL;
  size_t len = 0;

  while((getline(&line, &len, fp)) != -1) {
    buf_put(buf, line[0]);
  }
}

/**
 * Ecrit un buffer dans un fichier
 *
 * @param[out] fp  fichier de sortie
 * @param[in]  buf fichier d'entrée
 */
void writeBufferInFile(FILE* fp, struct buf* buf) {
  for(int i=0; i<buf->c; i++)
    fprintf(fp, "%c\n", buf->v[i]);
}

/**
 * Natual join
 *
 * @param[in]  buf_a   relation externe
 * @param[in]  buf_b   relation interne
 * @param[out] buf_out resultat du natural join de buf_a et buf_b
 */
void natural_join(struct buf* buf_a, struct buf* buf_b, struct buf* buf_out) {
  for(int a=0; a< buf_a->c; a++) {
    for(int b=0; b< buf_b->c; b++) {
      if(buf_a->v[a] == buf_b->v[b]) {
        buf_put(buf_out, buf_a->v[a]);
        break; // pass to next a
      }
    }
  }
}

/**
 * @brief Merge join
 *
 * @param[in]  buf_a   relation externe
 * @param[in]  buf_b   relation interne
 * @param[out] buf_out resultat du merge_join
 */
void merge_join(struct buf* buf_a, struct buf* buf_b, struct buf* buf_out) {
  char* buf_a_ptr= buf_a->v;
  char* buf_b_ptr= buf_b->v;

  // calculate pointer limit
  const char* buf_a_ptr_limit= buf_a->v + buf_a->c;
  const char* buf_b_ptr_limit= buf_b->v + buf_b->c;

  while( buf_a_ptr < buf_a_ptr_limit && buf_b_ptr_limit)
  {
    if(*buf_a_ptr == *buf_b_ptr) {
      buf_put(buf_out, *buf_a_ptr);
      buf_a_ptr++;
    } else if (*buf_a_ptr < *buf_b_ptr) {
      buf_a_ptr++;
    } else {
      buf_b_ptr++;
    }
  }
}

/**
 * Quick sort buffer
 */
void buf_quicksort(struct buf* buf) {
  quicksort(buf->v, 0, (buf->c) -1);
}
