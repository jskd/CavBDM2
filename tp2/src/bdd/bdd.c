/**
* TP n°: 2
*
* Titre du TP : Merge Join
*
* Date : 13/10/17
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques :
*
* History: 8871b16 [PASS]
*          5b1768f [PASS]
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "bdd.h"
#include "hexdump.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/**
 * @brief Structure de donnée repésentant un buffer
 */
struct buf {
  char* v;  //< Valeur
  size_t s; //< Taille
  size_t c; //< Nombre de valeur entrée
};

/**
 * Creation d'un buffer
 * @param[in] size Taille du buffer
 */
struct buf* buf_create( size_t size ) {
  struct buf* buf= (struct buf*) malloc(sizeof(struct buf));
  buf->v= (char*) malloc(size);
  buf->s= size;
  buf->c=0;
  return buf;
}

/**
 * Détruit le buffer
 * @param[in] buf buffer à détruire
 */
void buf_destroy(struct buf* buf) {
  free(buf->v);
  free(buf);
}

/**
 * Ajoute un caractere dans le buffer s'il reste de la place
 * @param[in] buf   buffer d'entrée
 * @param[in] value valeur à entrer
 * @return    0   succès
 *            -1  erreur: buffer plein
 */
char buf_put(struct buf* buf, char value) {
  if(buf->c >= buf->s)
    return -1;

  buf->v[buf->c]= value;
  buf->c++;

  return 0;
}

/**
 * @brief Enregistre un le 1er caractére de chaque
 *        ligne d'un fichier dans un buffer
 *
 * @param[in]  Fichier d'entrée
 * @param[out] Buffer de sortie
 */
void storeFileBuffer(FILE* fp, struct buf* buf) {
  char* line = NULL;
  size_t len = 0;

  while((getline(&line, &len, fp)) != -1) {
    buf_put(buf, line[0]);
  }
}

/**
 * @brief Ouvre un fichier
 *        Stoque son contenu dans un buffer
 *        Ferme le fichier
 *
 * @param[in]  File_name Fichier
 * @param[in] buffer_size Taille du buffer,
 *             si trop petit pour contenir le fichier entirerement
 *             alors le reste du fichier est ignoré
 *
 * @return buffer si NULL alors erreur de lecture du fichier
 *
 */
struct buf* storeFileBufferOC(const char* file_name, size_t buffer_size) {
  FILE* file = fopen(file_name, "r");
  if (file == NULL)
    return NULL;

  struct buf* buf= buf_create( buffer_size );
  storeFileBuffer(file, buf);
  fclose(file);

  return buf;
}

/**
 * Ecrit un buffer dans un fichier
 *
 * @param[out] fp  fichier de sortie
 * @param[in]  buf fichier d'entrée
 */
void writeBufferInFile(FILE* fp, const struct buf* buf) {
  for(int i=0; i<buf->c; i++)
    fprintf(fp, "%c\n", buf->v[i]);
}

/**
 * @brief Ouvre un fichier
 *        Ecrit le contenu du buffer dans le fichier
 *        Ferme le fichier
 *
 * @param[in]  File_name Fichier
 * @param[in]  buf       Buffer à écrire dans le fichier
 *
 * @return -1 si erreur dans l'ouverture du fichier
 *
 */
char writeBufferInFileOC(const char* file_name, const struct buf* buf) {

  FILE* file = fopen( file_name, "w");
  if (file == NULL)
    return -1;

  writeBufferInFile(file, buf);
  fclose(file);
  return 0;
}

/**
 * Natual join
 *
 * @param[in]  buf_a   relation externe
 * @param[in]  buf_b   relation interne
 * @param[out] buf_out resultat du natural join de buf_a et buf_b
 */
void natural_join(const struct buf* buf_a, const struct buf* buf_b, struct buf* buf_out) {
  for(int a=0; a< buf_a->c; a++) {
    for(int b=0; b< buf_b->c; b++) {
      if(buf_a->v[a] == buf_b->v[b]) {
        buf_put(buf_out, buf_a->v[a]);
        break; // pass to next a
      }
    }
  }
}

/**
 * @brief Merge join
 *
 * @param[in]  buf_a   relation a
 * @param[in]  buf_b   relation b
 * @param[out] buf_out resultat du merge_join
 */
void merge_join(const struct buf* buf_a, const struct buf* buf_b, struct buf* buf_out) {
  char* buf_a_ptr= buf_a->v; // pointer vers le buffer a
  char* buf_b_ptr= buf_b->v; // pointer vers le buffer b

  // Calcule la dernier case memoire du buffer + 1
  const char* buf_a_ptr_limit= buf_a->v + buf_a->c;
  const char* buf_b_ptr_limit= buf_b->v + buf_b->c;

  while( buf_a_ptr < buf_a_ptr_limit && buf_b_ptr < buf_b_ptr_limit)
  {
    if(*buf_a_ptr == *buf_b_ptr) {
      buf_put(buf_out, *buf_a_ptr);
      buf_a_ptr++;
    } else if (*buf_a_ptr < *buf_b_ptr) {
      buf_a_ptr++;
    } else {
      buf_b_ptr++;
    }
  }
}

/**
 * @brief Quick sort buffer
 * @param[in] buf buffer à trier
 */
void buf_quicksort(struct buf* buf) {
  quicksort(buf->v, 0, (buf->c) -1);
}

/**
 * @brief Hex dump buffer
 * @param[in] buf buffer à trier
 */
void buf_dump(const struct buf* buf) {
  printf ("buffer count: %d\n", buf->c);
  printf ("buffer size: %d\n", buf->s);
  hexDump("buffer value:", buf->v, (buf->s) -1);
}
