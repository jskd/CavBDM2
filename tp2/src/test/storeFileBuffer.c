/**
* TP n°: 1
*
* Titre du TP : Join Array
*
* Date : 06/10/17
*
* Nom : Lefranc
* Prenom : Joaquim
* email : lefrancjoaquim@gmail.com
*
* Nom : Skoda
* Prenom : Jérôme
* email : contact@jeromeskoda.fr
*
* Remarques : Code Block -> Trash. Makefile <3
*/

#include "../bdd/bdd.h"

static const size_t buf_size= 16;

int main(int argc, char** argv){

  // buffer R.txt
  struct buf* buf_r= storeFileBufferOC("res/R.txt", buf_size);
  if(buf_r == NULL) {
    printf("Erreur lors de la lecture de R.txt.\n");
    return -1;
  }

  buf_dump(buf_r);

  buf_destroy(buf_r);
  return 0;
}
