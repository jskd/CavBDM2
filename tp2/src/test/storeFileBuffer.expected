buffer count: 10
buffer size: 16
buffer value::
  0000  41 5a 47 4a 55 4b 45 42 56 44 00 00 00 00 00     AZGJUKEBVD.....
